plugin/v2/plugin.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Implementation of the KMS Plugin API v2.
+package v2
+
+import (
+	"regexp"
+
+	"google.golang.org/api/cloudkms/v1"
+
+	"context"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-cloudkms-plugin/plugin"
+	"github.com/golang/glog"
+)
+
+const (
+	apiVersion      = "v2beta1"
+	ok              = "ok"
+	ping            = "cGluZw=="
+	keyNotReachable = "Cloud KMS key is not reachable"
+	keyDisabled     = "Cloud KMS key is not enabled or no cloudkms.cryptoKeys.get permission"
+)
+
+// Store the last known primary key version resource name
+// to return as KeyId in case when the Cloud KMS service is not reachable
+// because KeyId in StatusResponse cannot be empty and shouldn't trigger key migration
+// in case of transient remote service unavailability
+var lastKeyId string
```

Comment:
This should be be scoped to the plugin instance, and may need a lock to protect against concurrent access.

Reply:
There is only one process/thread updating and reading the key, as I undersand. Adding concurrency check might unnecessary complicate the implementation.

Reply:
Probably better to say, that concurrent access can only occur in case Kubernetes calls Status and Encrypt/Decrypt concurrently/simultaneousely. I don't think that that would work reliably on the Kubernetes side if they would be doing that. Hence I am relying here on Kubernetes tracking its KMS API calls concurency.